src/Models/UserIdentityModel.php

```diff
@@ -25,6 +26,146 @@ class UserIdentityModel extends Model
+    {
+        return $this
+            ->where('type', 'access_token')
+            ->where('secret', hash('sha256', $token))
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * @param int|string $userId
+     */
+    public function getAccessToken($userId, string $token): ?AccessToken
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->where('secret', hash('sha256', $token))
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * Given the ID, returns the given access token.
+     *
+     * @param int|string $id
+     * @param int|string $userId
+     */
+    public function getAccessTokenById($id, $userId): ?AccessToken
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->where('id', $id)
+            ->asObject(AccessToken::class)
+            ->first();
+    }
+
+    /**
+     * @param int|string $userId
+     *
+     * @return AccessToken[]
+     */
+    public function getAllAccessTokens($userId): array
+    {
+        return $this
+            ->where('user_id', $userId)
+            ->where('type', 'access_token')
+            ->asObject(AccessToken::class)
+            ->find();
+    }
+
+    /**
+     * @param int|string $userId
+     */
+    public function getIdentityByType($userId, string $type): ?UserIdentity
+    {
+        return $this->where('user_id', $userId)
+            ->where('type', $type)
+            ->first();
+    }
+
+    public function getIdentityBySecret(string $type, ?string $secret): ?UserIdentity
+    {
+        if ($secret === null) {
+            return null;
+        }
+
+        return $this->where('type', $type)
+            ->where('secret', $secret)
+            ->first();
+    }
+
+    /**
+     * @param int|string $userId
+     *
+     * @return UserIdentity[]
+     */
+    public function getIdentities($userId): array
+    {
+        return $this->where('user_id', $userId)->findAll();
+    }
+
+    /**
+     * @param int[]|string[] $userIds
+     *
+     * @return UserIdentity[]|null
+     */
+    public function getIdentitiesByUserIds(array $userIds)
+    {
+        return $this->whereIn('user_id', $userIds)->find();
+    }
+
+    /**
+     * @param int|string $userId
+     * @param string[]   $types
+     *
+     * @return UserIdentity[]
+     */
+    public function getIdentitiesByTypes($userId, array $types): array
+    {
+        return $this->where('user_id', $userId)
+            ->whereIn('type', $types)
+            ->findAll();
+    }
+
+    /**
+     * @param int|string $userId
+     */
+    public function deleteIdentitiesByType($userId, string $type): void
+    {
+        $this->where('user_id', $userId)
+            ->where('type', $type)
+            ->delete();
+    }
+
+    /**
+     * Given the token, will retrieve the token to
+     * verify it exists, then delete it.
```

Comment:
More confusion over "token". But also, this appears to delete blindly without the "verify it exists" part.

Reply:
I am not sure what to do when it does not exist.

Reply:
I think your solution is fine, there's a lot of precedence for delete operations being idempotent - I would just update the comment:
```php
     * Delete any access tokens for the given rawToken
```

Reply:
Done!